Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/SubmissionsController.cs

```diff
@@ -330,37 +336,70 @@ public JsonResult GetSubmissionTypes(int problemId, bool? allowBinaryFilesUpload
             return this.Json(submissionTypesSelectListItems, JsonRequestBehavior.AllowGet);
         }
 
+        [HttpPost]
+        [ValidateAntiForgeryToken]
+        [AllowAnonymous]
+        [RestrictRequests(
+            RequestsPerInterval = RequestsPerInterval,
+            RestrictInterval = RestrictInterval,
+            ErrorMessage = TooManyRequestsErrorMessage)]
         public ActionResult Retest(int id)
         {
             var submission = this.Data.Submissions.GetById(id);
 
+            if (!this.ModelState.IsValid)
+            {
+                var modelStateErrors = this.ModelState.Values.SelectMany(m => m.Errors);
+                foreach (var modelStateError in modelStateErrors)
+                {
+                    this.TempData.AddDangerMessage(modelStateError.ErrorMessage);
+                }
+
+                return this.RedirectToAction("Index", "Contests", new {area = ""});
+            }
+
             if (submission == null)
             {
                 this.TempData.AddDangerMessage(Resource.Invalid_submission_message);
+                return this.RedirectToAction("Index", "Contests", new { area = "" });
             }
-            else
+
+            bool isAdmin = this.User.IsAdmin();
+            if (isAdmin)
```

Comment:
there is no need of different var here. You can move `this.User.IsAdmin()` in `if`